src/cargo/core/compiler/mod.rs

```diff
@@ -629,19 +631,42 @@ fn rustdoc<'a, 'cfg>(cx: &mut Context<'a, 'cfg>, unit: &Unit<'a>) -> CargoResult
     }))
 }
 
-fn add_crate_versions_if_requested(
-    bcx: &BuildContext<'_, '_>,
-    unit: &Unit<'_>,
+fn add_crate_versions_if_requested<'a>(
+    bcx: &BuildContext<'a, '_>,
+    unit: &Unit<'a>,
     rustdoc: &mut ProcessBuilder,
 ) {
-    if !bcx.config.cli_unstable().crate_versions {
-        return;
+    if crate_versions_requested(bcx)
+        && !crate_version_flag_in_rustdocflags(bcx, unit)
+        && !crate_version_flag_in_extra_args(bcx, unit)
+    {
+        append_crate_version_flag(unit, rustdoc);
     }
+}
+
+fn crate_versions_requested(bcx: &BuildContext<'_, '_>) -> bool {
+    bcx.config.cli_unstable().crate_versions
+}
```

Comment:
I don't think this necessarily needs to be a separate function, I think it would be best just to put the check inline.  I can appreciate having the function names clearly express what is being done, but wrapping single expressions like this tends to significantly increase the total lines of code.